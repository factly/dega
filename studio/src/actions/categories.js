import axios from 'axios';
import {
  ADD_CATEGORIES,
  ADD_CATEGORIES_REQUEST,
  SET_CATEGORIES_LOADING,
  RESET_CATEGORIES,
  CATEGORIES_API,
  GET_CATEGORY,
  UPDATE_CATEGORY,
} from '../constants/categories';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addMedia } from './media';
import getError from '../utils/getError';

// action to fetch all categories
export const getCategories = (query) => {
  return (dispatch, getState) => {
    const currentSpaceID = getState().spaces?.selected;
    if (currentSpaceID === 0) {
      return;
    }
    dispatch(loadingCategories());
    return axios
      .get(CATEGORIES_API, {
        params: query,
      })
      .then((response) => {
        dispatch(
          addMedia(
            response.data.nodes
              .filter((category) => category.medium)
              .map((category) => category.medium),
          ),
        );
        dispatch(
          addCategoriesList(
            response.data.nodes.map((category) => {
              return { ...category, medium: category.medium?.id };
            }),
          ),
        );
        dispatch(
          addCategoriesRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to fetch category by id
export const getCategory = (id) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .get(CATEGORIES_API + '/' + id)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));

        dispatch(addCategory(GET_CATEGORY, { ...response.data, medium: response.data.medium?.id }));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to create category
export const createCategory = (data) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .post(CATEGORIES_API, data)
      .then(() => {
        dispatch(resetCategories());
        dispatch(addSuccessNotification('Category created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update category by id
export const updateCategory = (data) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .put(CATEGORIES_API + '/' + data.id, data)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));

        dispatch(
          addCategory(UPDATE_CATEGORY, { ...response.data, medium: response.data.medium?.id }),
        );
        dispatch(addSuccessNotification('Category updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to delete category by id
export const deleteCategory = (id) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .delete(CATEGORIES_API + '/' + id)
      .then(() => {
        dispatch(resetCategories());
        dispatch(addSuccessNotification('Category deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addCategories = (categories) => {
  return (dispatch) => {
    dispatch(
      addMedia(categories.filter((category) => category.medium).map((category) => category.medium)),
    );
    dispatch(
      addCategoriesList(
        categories.map((category) => {
          return { ...category, medium: category.medium?.id };
        }),
      ),
    );
  };
};

export const loadingCategories = () => ({
  type: SET_CATEGORIES_LOADING,
  payload: true,
});

export const stopCategoriesLoading = () => ({
  type: SET_CATEGORIES_LOADING,
  payload: false,
});

export const addCategory = (type, payload) => ({
  type,
  payload,
});

export const addCategoriesList = (data) => ({
  type: ADD_CATEGORIES,
  payload: data,
});

export const addCategoriesRequest = (data) => ({
  type: ADD_CATEGORIES_REQUEST,
  payload: data,
});

export const resetCategories = () => ({
  type: RESET_CATEGORIES,
});
