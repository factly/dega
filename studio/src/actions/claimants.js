import axios from 'axios';
import {
  ADD_CLAIMANTS,
  ADD_CLAIMANTS_REQUEST,
  SET_CLAIMANTS_LOADING,
  RESET_CLAIMANTS,
  CLAIMANTS_API,
  UPDATE_CLAIMANT,
  GET_CLAIMANT,
} from '../constants/claimants';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addMedia } from './media';
import getError from '../utils/getError';

// action to fetch all claimants
export const getClaimants = (query) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .get(CLAIMANTS_API, {
        params: query,
      })
      .then((response) => {
        dispatch(
          addMedia(
            response.data.nodes
              .filter((claimant) => claimant.medium)
              .map((claimant) => claimant.medium),
          ),
        );
        dispatch(
          addClaimantsList(
            response.data.nodes.map((claimant) => {
              claimant.description = {
                json: claimant.description,
                html: claimant.html_description,
              };
              return { ...claimant, medium: claimant.medium?.id };
            }),
          ),
        );
        dispatch(
          addClaimantsRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to fetch claimant by id
export const getClaimant = (id) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .get(CLAIMANTS_API + '/' + id)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));
        response.data.description = {
          json: response.data.description,
          html: response.data.html_description,
        };
        dispatch(addClaimant(GET_CLAIMANT, { ...response.data, medium: response.data.medium?.id }));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to create claimant
export const createClaimant = (data) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .post(CLAIMANTS_API, data)
      .then(() => {
        dispatch(resetClaimants());
        dispatch(addSuccessNotification('Claimant created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update claimant by id
export const updateClaimant = (data) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .put(CLAIMANTS_API + '/' + data.id, data)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));
        response.data.description = {
          json: response.data.description,
          html: response.data.html_description,
        };
        dispatch(
          addClaimant(UPDATE_CLAIMANT, { ...response.data, medium: response.data.medium?.id }),
        );
        dispatch(addSuccessNotification('Claimant updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to delete claimant by id
export const deleteClaimant = (id) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .delete(CLAIMANTS_API + '/' + id)
      .then(() => {
        dispatch(resetClaimants());
        dispatch(addSuccessNotification('Claimant deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addClaimants = (claimants) => {
  return (dispatch) => {
    dispatch(
      addMedia(claimants.filter((claimant) => claimant.medium).map((claimant) => claimant.medium)),
    );
    dispatch(
      addClaimantsList(
        claimants.map((claimant) => {
          return { ...claimant, medium: claimant.medium?.id };
        }),
      ),
    );
  };
};

export const loadingClaimants = () => ({
  type: SET_CLAIMANTS_LOADING,
  payload: true,
});

export const stopClaimantsLoading = () => ({
  type: SET_CLAIMANTS_LOADING,
  payload: false,
});

export const addClaimant = (type, payload) => ({
  type,
  payload,
});

export const addClaimantsList = (data) => ({
  type: ADD_CLAIMANTS,
  payload: data,
});

export const addClaimantsRequest = (data) => ({
  type: ADD_CLAIMANTS_REQUEST,
  payload: data,
});

export const resetClaimants = () => ({
  type: RESET_CLAIMANTS,
});
