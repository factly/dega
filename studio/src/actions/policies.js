import axios from 'axios';
import {
  ADD_POLICIES,
  ADD_POLICIES_REQUEST,
  SET_POLICIES_LOADING,
  RESET_POLICIES,
  POLICIES_API,
  UPDATE_POLICY,
  GET_POLICY,
} from '../constants/policies';
import { addErrorNotification, addSuccessNotification } from './notifications';
import getError from '../utils/getError';

// action to fetch all policies
export const addDefaultPolicies = (query) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .post(POLICIES_API + '/default')
      .then((response) => {
        dispatch(addPolicies(response.data.nodes));
        dispatch(
          addPoliciesRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopPoliciesLoading()));
  };
};

// action to fetch all policies
export const getPolicies = (query) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .get(POLICIES_API, {
        params: query,
      })
      .then((response) => {
        dispatch(addPolicies(response.data.nodes));
        dispatch(
          addPoliciesRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopPoliciesLoading()));
  };
};

// action to fetch policy by id
export const getPolicy = (id) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .get(POLICIES_API + '/' + id)
      .then((response) => {
        dispatch(addPolicy(GET_POLICY, response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopPoliciesLoading()));
  };
};

// action to create policy
export const createPolicy = (data) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .post(POLICIES_API, data)
      .then(() => {
        dispatch(resetPolicies());
        dispatch(addSuccessNotification('Policy created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update policy by id
export const updatePolicy = (data) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .put(POLICIES_API + '/' + data.id, data)
      .then((response) => {
        dispatch(addPolicy(UPDATE_POLICY, response.data));
        dispatch(addSuccessNotification('Policy updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopPoliciesLoading()));
  };
};

// action to delete policy by id
export const deletePolicy = (id) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .delete(POLICIES_API + '/' + id)
      .then(() => {
        dispatch(resetPolicies());
        dispatch(addSuccessNotification('Policy deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const loadingPolicies = () => ({
  type: SET_POLICIES_LOADING,
  payload: true,
});

export const stopPoliciesLoading = () => ({
  type: SET_POLICIES_LOADING,
  payload: false,
});

export const addPolicy = (type, payload) => ({
  type,
  payload,
});

export const addPolicies = (payload) => ({
  type: ADD_POLICIES,
  payload,
});

export const addPoliciesRequest = (payload) => ({
  type: ADD_POLICIES_REQUEST,
  payload,
});

export const resetPolicies = () => ({
  type: RESET_POLICIES,
});
