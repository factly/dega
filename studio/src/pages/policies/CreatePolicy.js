import React from 'react';
import { createPolicy } from '../../actions/policies';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import PolicyCreateForm from './components/PolicyForm';
import { Helmet } from 'react-helmet';

function CreatePolicy() {
  const history = useHistory();

  const dispatch = useDispatch();
  const onCreate = (values) => {
    values.permissions = values.permissions?.filter(
      (item) => item && item.resource && item.actions.length > 0,
    );
    dispatch(createPolicy(values)).then(() => history.push('/members/policies'));
  };

  return (
    <>
      <Helmet title={'Create Policy'} />
      <PolicyCreateForm onCreate={onCreate} />
    </>
  );
}

export default CreatePolicy;
