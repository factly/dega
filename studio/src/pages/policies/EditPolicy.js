import React from 'react';
import PolicyEditForm from './components/PolicyForm';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';
import { getPolicy, updatePolicy } from '../../actions/policies';
import { useHistory } from 'react-router-dom';
import { useParams } from 'react-router-dom';
import RecordNotFound from '../../components/ErrorsAndImage/RecordNotFound';
import { Helmet } from 'react-helmet';

function EditPolicy() {
  const history = useHistory();
  const { id } = useParams();

  const dispatch = useDispatch();

  const { policy, loading } = useSelector((state) => {
    if (!state.policies.details[id])
      return {
        policy: null,
        loading: state.policies.loading,
      };

    return {
      policy: {
        ...state.policies.details[id],
        permissions: state.policies.details[id].permissions.reduce(
          (obj, item) => Object.assign(obj, { [item.resource]: item.actions }),
          {},
        ),
        roles: state.policies.details[id].roles?.length 
          ? state.policies.details[id].roles.map((item) => item.id)
          : [],
      },
      loading: state.policies.loading,
    };
  });

  React.useEffect(() => {
    dispatch(getPolicy(id));
  }, [dispatch, id]);

  if (loading) return <Skeleton />;

  if (!policy) {
    return <RecordNotFound />;
  }

  const onUpdate = (values) => {
    dispatch(updatePolicy({ ...policy, ...values })).then(() =>
      history.push(`/members/policies/`),
    );
  };

  return (
    <>
      <Helmet title={`${policy.name} - Edit Policy`} />
      <PolicyEditForm data={policy} onCreate={onUpdate} />
    </>
  );
}

export default EditPolicy;
