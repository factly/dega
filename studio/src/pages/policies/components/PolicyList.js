import React from 'react';
import { Modal, Button, Typography, Table, ConfigProvider } from 'antd';

import { useDispatch, useSelector } from 'react-redux';
import { deletePolicy } from '../../../actions/policies';
import { Link, useHistory } from 'react-router-dom';
import { DeleteOutlined, EyeOutlined } from '@ant-design/icons';

function PolicyList({ actions, data, filters, setFilters, fetchPolicies }) {
  const [modalOpen, setModalOpen] = React.useState(false);

  const dispatch = useDispatch();

  const history = useHistory();

  const columns = [
    {
      title: 'Name',
      dataIndex: 'name',
      key: 'name',
      width: 200,
      onCell: () => {
        return {
          style: {
            minWidth: '200px',
          },
        };
      },
      render: (_, record) => {
        return (
          <Link
            style={{
              marginRight: 8,
            }}
            to={`settings/members/policies/${record.id}/edit`}
          >
            <Typography.Text style={{ fontSize: '1rem' }} strong>
              {record.name}
            </Typography.Text>
          </Link>
        );
      },
    },
    {
      title: 'Description',
      dataIndex: 'description',
      key: 'description',
      width: 400,
      onCell: () => {
        return {
          style: {
            minWidth: '400px',
          },
        };
      },
      render: (_, record) => {
        return (
          <Typography.Paragraph style={{ fontSize: '1rem' }} strong ellipsis={{ rows: 2 }}>
            {record.description}
          </Typography.Paragraph>
        );
      },
    },
    {
      title: 'Action',
      dataIndex: 'operation',
      align: 'center',
      width: 200,
      onCell: () => {
        return {
          style: {
            minWidth: '200px',
          },
        };
      },
      render: (_, record) => {
        return (
          <ConfigProvider
            theme={{
              components: {
                Button: {
                  controlHeight: 35,
                  colorBorder: '#F2F2F2',
                  colorPrimaryHover: '#00000026',
                },
                Modal: {
                  colorBgMask: '#0000000B',
                },
              },
            }}
          >
            <Link
              to={{
                pathname: `/settings/members/policies/${record.id}/view`,
              }}
              onClick={(e) => {
                e.stopPropagation();
              }}
            >
              <Button
                icon={<EyeOutlined style={{ color: '#858585' }} />}
                size="large"
                style={{
                  marginRight: 8,
                }}
              />
            </Link>
            <Button
              size="large"
              onClick={(e) => {
                e.stopPropagation();
                setModalOpen(true);
              }}
              icon={<DeleteOutlined style={{ color: '#858585' }} />}
              disabled={!(actions.includes('admin') || actions.includes('delete'))}
            />
            <Modal
              open={modalOpen}
              closable={false}
              centered
              width={311}
              className="delete-modal-container"
              style={{
                borderRadius: '18px',
              }}
              onOk={(e) => {
                e.stopPropagation();
                dispatch(deletePolicy(record.id)).then(() => fetchPolicies());
                setModalOpen(false);
              }}
              disabled={!(actions.includes('admin') || actions.includes('delete'))}
              cancelButtonProps={{ type: 'text', style: { color: '#000' } }}
              onCancel={(e) => {
                e.stopPropagation();
                setModalOpen(false);
              }}
            >
              <Typography.Text strong>Are you sure you want to delete this ?</Typography.Text>
            </Modal>
          </ConfigProvider>
        );
      },
    },
  ];

  return (
    <ConfigProvider
      theme={{
        components: {
          Typography: {
            colorText: '#101828',
          },
        },
      }}
    >
      <Table
        onRow={(record, rowIndex) => {
          return {
            onClick: (event) => {
              history.push(`/settings/members/policies/${record.id}/edit`);
            },
            onMouseEnter: (event) => {
              document.body.style.cursor = 'pointer';
            },
            onMouseLeave: (event) => {
              document.body.style.cursor = 'default';
            },
          };
        }}
        scroll={{
          x: '1000',
        }}
        columns={columns}
        dataSource={data.policies}
        loading={data.loading}
        rowKey={'name'}
        pagination={{
          showTotal: (total, range) => `${range[0]}-${range[1]} of ${total} results`,
          total: data.total,
          current: filters.page,
          pageSize: filters.limit,
          onChange: (pageNumber, pageSize) => setFilters({ page: pageNumber, limit: pageSize }),
          pageSizeOptions: ['10', '15', '20'],
        }}
      />
    </ConfigProvider>
  );
}

export default PolicyList;
