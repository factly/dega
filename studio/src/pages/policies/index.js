import React from 'react';
import { Space, Button, Row } from 'antd';
import { Link, useLocation } from 'react-router-dom';
import PolicyList from './components/PolicyList';
import getUserPermission from '../../utils/getUserPermission';
import { useDispatch, useSelector } from 'react-redux';
import { getPolicies } from '../../actions/policies';
import deepEqual from 'deep-equal';
import Loader from '../../components/Loader';
import { Helmet } from 'react-helmet';

function Policies() {
  const spaces = useSelector(({ spaces }) => spaces);
  const actions = getUserPermission({ resource: 'policies', action: 'get', spaces });
  const dispatch = useDispatch();
  const query = new URLSearchParams(useLocation().search);
  const [filters, setFilters] = React.useState({
    page: 1,
    limit: 20,
  });
  query.set('page', filters.page);
  window.history.replaceState({}, '', `${window.PUBLIC_URL}${useLocation().pathname}?${query}`);
  const { policies, total, loading } = useSelector((state) => {
    const node = state.policies.req.find((item) => {
      return deepEqual(item.query, filters);
    });

    if (node)
      return {
        policies: node.data.map((element) => state.policies.details[element]),
        total: node.total,
        loading: state.policies.loading,
      };
    return { policies: [], total: 0, loading: state.policies.loading };
  });

  React.useEffect(() => {
    fetchPolicies();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [filters]);

  const fetchPolicies = () => {
    dispatch(getPolicies(filters));
  };

  return loading ? (
    <Loader />
  ) : (
    <Space direction="vertical">
      <Helmet title={'Policies'} />
      <Row gutter={16} justify="end">
        <Link to="/members/policies/create">
          <Button
            disabled={!(actions.includes('admin') || actions.includes('create'))}
            type="primary"
          >
            New Policy
          </Button>
        </Link>
      </Row>
      <PolicyList
        actions={actions}
        data={{ policies, total, loading }}
        filters={filters}
        setFilters={setFilters}
        fetchPolicies={fetchPolicies}
      />
    </Space>
  );
}

export default Policies;
