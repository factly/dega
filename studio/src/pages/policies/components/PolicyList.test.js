import React from 'react';
import { BrowserRouter as Router, Link } from 'react-router-dom';

import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Popconfirm, Button, Table } from 'antd';

import '../../../matchMedia.mock';
import PolicyList from './PolicyList';
import { getPolicies, deletePolicy } from '../../../actions/policies';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
let mockedDispatch, store;
const setFilters = jest.fn();
const fetchPolicies = jest.fn();
const filters = {
  page: 1,
  limit: 20,
};
let info = {
  policies: [
    {
      id: '1',
      name: 'Test',
      description: '',
      permissions: [
        {
          resource: 'tags',
          actions: ['get'],
        },
      ],
      users: [
        {
          id: 1,
          created_at: '2020-09-23T11:06:11.308302Z',
          updated_at: '2020-09-23T11:36:09.244259Z',
          deleted_at: null,
          email: 'ross@gmail.com',
          first_name: 'Ross',
          last_name: 'Geller',
          birth_date: '1997-12-23T17:05:55+05:30',
          gender: 'male',
        },
      ],
    },
  ],
  total: 1,
  loading: false,
};
let state = {
  policies: {
    req: [
      {
        data: ['1', '2'],
        query: {
          page: 1,
          limit: 20,
        },
        total: 2,
      },
    ],
    details: {
      1: {
        id: '1',
        name: 'Test',
        description: '',
        permissions: [
          {
            resource: 'tags',
            actions: ['get'],
          },
        ],
        users: [
          {
            id: 1,
            created_at: '2020-09-23T11:06:11.308302Z',
            updated_at: '2020-09-23T11:36:09.244259Z',
            deleted_at: null,
            email: 'ross@gmail.com',
            first_name: 'Ross',
            last_name: 'Geller',
            birth_date: '1997-12-23T17:05:55+05:30',
            gender: 'male',
          },
        ],
      },
      2: {
        id: '2',
        name: 'Test-2',
        description: '',
        permissions: [
          {
            resource: 'categories',
            actions: ['get'],
          },
        ],
        users: [
          {
            id: 2,
            created_at: '2020-10-23T11:06:11.308302Z',
            updated_at: '2020-10-23T11:36:09.244259Z',
            deleted_at: null,
            email: 'ross2@gmail.com',
            first_name: 'Ross',
            last_name: 'Geller',
            birth_date: '1995-12-23T17:05:55+05:30',
            gender: 'male',
          },
        ],
      },
    },
    loading: false,
  },
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));
jest.mock('../../../actions/policies', () => ({
  getPolicies: jest.fn(),
  deletePolicy: jest.fn(),
}));

describe('Policies List component', () => {
  describe('snapshot testing', () => {
    beforeEach(() => {
      store = mockStore({});
      store.dispatch = jest.fn();
      mockedDispatch = jest.fn();
      useDispatch.mockReturnValue(mockedDispatch);
    });
    it('should render the component', () => {
      store = mockStore(state);
      const tree = shallow(
        <Provider store={store}>
          <Router>
            <PolicyList actions={['update', 'delete']} />
          </Router>
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
    it('should match component when loading', () => {
      state.policies.loading = true;
      store = mockStore(state);
      const tree = shallow(
        <Provider store={store}>
          <Router>
            <PolicyList actions={['update', 'delete']} />
          </Router>
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
    it('should match component with policies', () => {
      state.policies.loading = false;
      store = mockStore(state);
      const tree = shallow(
        <Provider store={store}>
          <Router>
            <PolicyList actions={['update', 'delete']} />
          </Router>
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      mockedDispatch = jest.fn(() => new Promise((resolve) => resolve(true)));
      useDispatch.mockReturnValue(mockedDispatch);
    });
    it('should change the page', () => {
      store = mockStore(state);
      let wrapper;
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <Router>
              <PolicyList
                setFilters={setFilters}
                fetchPolicies={fetchPolicies}
                filters={filters}
                data={info}
                actions={['update', 'delete']}
              />
            </Router>
          </Provider>,
        );
      });

      const table = wrapper.find(Table);
      table.props().pagination.onChange(1);
      wrapper.update();
      const updatedTable = wrapper.find(Table);
      expect(updatedTable.props().pagination.current).toEqual(1);
    });
    it('should delete policy', () => {
      store = mockStore(state);
      let wrapper;
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <Router>
              <PolicyList
                setFilters={setFilters}
                fetchPolicies={fetchPolicies}
                filters={filters}
                data={info}
                actions={['update', 'delete']}
              />
            </Router>
          </Provider>,
        );
      });
      const button = wrapper.find(Button).at(0);
      expect(button.text()).toEqual('');

      button.simulate('click');
      const popconfirm = wrapper.find(Popconfirm);
      popconfirm
        .findWhere((item) => item.type() === 'button' && item.text() === 'OK')
        .simulate('click');
      expect(deletePolicy).toHaveBeenCalled();
      expect(deletePolicy).toHaveBeenCalledWith('1');
      //  expect(getPolicies).toHaveBeenCalledWith({ page: 1, limit: 20 });
    });
    it('should edit policy', () => {
      store = mockStore(state);
      let wrapper;
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <Router>
              <PolicyList
                setFilters={setFilters}
                fetchPolicies={fetchPolicies}
                filters={filters}
                data={info}
                actions={['update', 'delete']}
              />
            </Router>
          </Provider>,
        );
      });
      const link = wrapper.find(Link).at(0);
      expect(link.text()).toEqual('Test');
      expect(link.prop('to')).toEqual('/members/policies/1/edit');
    });
    it('should have no delete and edit buttons', () => {
      store = mockStore({
        policies: {
          req: [],
        },
      });
      let wrapper;
      const info2 = { ...info };
      info2.policies = [];
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <Router>
              <PolicyList
                setFilters={setFilters}
                fetchPolicies={fetchPolicies}
                filters={filters}
                data={info2}
                actions={['update', 'delete']}
              />
            </Router>
          </Provider>,
        );
      });

      const button = wrapper.find(Button);
      expect(button.length).toEqual(0);
    });
  });
});
