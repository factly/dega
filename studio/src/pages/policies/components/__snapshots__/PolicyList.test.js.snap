// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policies List component snapshot testing should match component when loading 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <PolicyList
      actions={
        Array [
          "update",
          "delete",
        ]
      }
    />
  </BrowserRouter>
</ContextProvider>
`;

exports[`Policies List component snapshot testing should match component with policies 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <PolicyList
      actions={
        Array [
          "update",
          "delete",
        ]
      }
    />
  </BrowserRouter>
</ContextProvider>
`;

exports[`Policies List component snapshot testing should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <PolicyList
      actions={
        Array [
          "update",
          "delete",
        ]
      }
    />
  </BrowserRouter>
</ContextProvider>
`;
